### PR TITLE
feat(cdn) - Feature parity

### PR DESCRIPTION
**Pull request by matheuspolitano**
## Description

(https://jira.schwarz/browse/STACKITCDN-1451)

## Checklist

- [x] Issue was linked above
- [x] Code format was applied: `make fmt`
- [x] Examples were added / adjusted (see `examples/` directory)
- [x] Docs are up-to-date: `make generate-docs` (will be checked by CI)
- [x] Unit tests got implemented or updated
- [x] Acceptance tests got implemented or updated (see e.g. [here](https://github.com/stackitcloud/terraform-provider-stackit/blob/f5f99d170996b208672ae684b6da53420e369563/stackit/internal/services/dns/dns_acc_test.go))
- [x] Unit tests are passing: `make test` (will be checked by CI)
- [x] No linter issues: `make lint` (will be checked by CI)

### CODE REVIEW

**Comment by cgoetz-inovex** [comment]
Hi @matheuspolitano,
thanks for your contribution, I've created an internal issue to review and test this PR.

File: stackit/internal/services/cdn/distribution/datasource_test.go [hunk 1]
```diff
@@ -66,6 +66,9 @@ func TestMapDataSourceFields(t *testing.T) {
 		"backend":                backend,
 		"regions":                regionsFixture,
 		"blocked_countries":      blockedCountriesFixture,
+		"blocked_ips":            types.ListValueMust(types.StringType, []attr.Value{}),
```
Comment (Manuelvaas) [inline stackit/internal/services/cdn/distribution/datasource_test.go:69]: Could you add some tests where the values are not null

File: stackit/internal/services/cdn/distribution/resource_test.go [hunk 2]
```diff
@@ -13,6 +13,19 @@ import (
 	cdnSdk "github.com/stackitcloud/stackit-sdk-go/services/cdn/v1api"
 )
 
+func createTestConfig(vals map[string]attr.Value) types.Object {
```
Comment (Manuelvaas) [inline stackit/internal/services/cdn/distribution/resource_test.go:16]: Please add tests here as well to test the values. The createTestConfig is kind of special. Maybe you could use a fixture config like in other tests instead of this?

File: stackit/internal/services/cdn/distribution/resource.go [hunk 3]
```diff
@@ -714,6 +723,23 @@ func (r *distributionResource) Schema(_ context.Context, _ resource.SchemaReques
 						// API response (empty list).
 						Default: listdefault.StaticValue(types.ListValueMust(types.StringType, []attr.Value{})),
 					},
+					"blocked_ips": schema.ListAttribute{
+						Optional:    true,
+						Computed:    true, // Required when using Default
+						Description: schemaDescriptions["config_blocked_ips"],
+						ElementType: types.StringType,
+						Default:     listdefault.StaticValue(types.ListValueMust(types.StringType, []attr.Value{})),
```
Comment (Manuelvaas) [inline stackit/internal/services/cdn/distribution/resource.go]: We try to avoid using defaults because then the value gets actively set by Terraform, and you cannot use a server-side default or change this later. Feel free to reach out to me if you have further questions regarding this.

File: stackit/internal/services/cdn/distribution/resource.go [hunk 4]
```diff
@@ -714,6 +723,23 @@ func (r *distributionResource) Schema(_ context.Context, _ resource.SchemaReques
...
+					},
+					"default_cache_duration": schema.StringAttribute{
+						Optional:    true,
+						Computed:    true,
```
Comment (Manuelvaas) [inline stackit/internal/services/cdn/distribution/resource.go:735]: Is this value changed by the service?

File: stackit/internal/services/cdn/distribution/resource.go [hunk 5]
```diff
@@ -714,6 +723,23 @@ func (r *distributionResource) Schema(_ context.Context, _ resource.SchemaReques
...
+						Description: schemaDescriptions["config_default_cache_duration"],
+					},
+					"monthly_limit_bytes": schema.Int64Attribute{
+						Optional:    true,
+						Computed:    true,
```
Comment (Manuelvaas) [inline stackit/internal/services/cdn/distribution/resource.go:740]: Is this value changed by the service?

File: stackit/internal/services/cdn/distribution/resource.go [hunk 6]
```diff
@@ -993,6 +1026,12 @@ func (r *distributionResource) Update(ctx context.Context, req resource.UpdateRe
 	if !utils.IsUndefined(configModel.StripResponseCookies) {
 		configPatch.StripResponseCookies = configModel.StripResponseCookies.ValueBoolPointer()
 	}
+	if !utils.IsUndefined(configModel.DefaultCacheDuration) {
```
Comment (Manuelvaas) [inline stackit/internal/services/cdn/distribution/resource.go:1033]: Interesting that there is a toCreatePayload but no toUpdatePayload function 🤔 If you like, you could introduce that, but it's not a must-have because it was already that way. But would be much cleaner and better testable.

**Review by Manuelvaas — CHANGES_REQUESTED** [review]

File: stackit/internal/services/cdn/distribution/resource.go (hunk 3 above)
Comment (matheuspolitano) [reply stackit/internal/services/cdn/distribution/resource.go]: If we don't define a schema Default, Terraform plans the omitted attribute as null. But when the API returns [], mapFields writes [] into the state. Because null (plan) does not match [] (state), Terraform throws the "Provider produced inconsistent result after apply" error.  Adding Default: listdefault.StaticValue(...) forces the planned value to be [] in the plan before apply, making it match the returned state perfectly and preventing the error.

Comment (matheuspolitano) [reply stackit/internal/services/cdn/distribution/resource.go]: Also this value will never change so is safe

File: stackit/internal/services/cdn/distribution/resource.go (hunk 5 above)
Comment (matheuspolitano) [reply stackit/internal/services/cdn/distribution/resource.go:740]: Yep

File: stackit/internal/services/cdn/distribution/resource.go (hunk 4 above)
Comment (matheuspolitano) [reply stackit/internal/services/cdn/distribution/resource.go:735]: Yep

File: stackit/internal/services/cdn/distribution/resource.go (hunk 6 above)
Comment (matheuspolitano) [reply stackit/internal/services/cdn/distribution/resource.go:1033]: We have a plan to refactor in the future

File: stackit/internal/services/cdn/distribution/resource_test.go (hunk 2 above)
Comment (matheuspolitano) [reply stackit/internal/services/cdn/distribution/resource_test.go:16]: Its done

File: stackit/internal/services/cdn/distribution/datasource_test.go (hunk 1 above)
Comment (matheuspolitano) [reply stackit/internal/services/cdn/distribution/datasource_test.go:69]: Done

File: stackit/internal/services/cdn/distribution/resource.go (hunk 3 above)
Comment (matheuspolitano) [reply stackit/internal/services/cdn/distribution/resource.go]: After your dm i've removed the default field

**Review by Manuelvaas — DISMISSED** [review]

File: stackit/internal/services/cdn/distribution/resource.go [hunk 7]
```diff
@@ -993,6 +1030,12 @@ func (r *distributionResource) Update(ctx context.Context, req resource.UpdateRe
...
+		configPatch.DefaultCacheDuration = *cdnSdk.NewNullableString(conversion.StringValueToPointer(configModel.DefaultCacheDuration))
+	}
+	if !utils.IsUndefined(configModel.MonthlyLimitBytes) {
+		configPatch.MonthlyLimitBytes = *cdnSdk.NewNullableInt64(conversion.Int64ValueToPointer(configModel.MonthlyLimitBytes))
```
Comment (cgoetz-inovex) [inline stackit/internal/services/cdn/distribution/resource.go:1037]: example config:
```hcl
   resource "stackit_cdn_distribution" "example-distribution" {
     project_id = var.project_id

     config = {
       backend = {
         type       = "http"
         origin_url = "https://example.com"
       }

       regions = ["EU"]

       default_cache_duration = "PT2H"
       monthly_limit_bytes    = 104857600
     }
   }
```
run `terraform apply` then remove `default_cache_duration` and `monthly_limit_bytes`.
Then run `apply` again.

snippet from state:
```
        default_cache_duration = "PT2H"
        forward_host_header    = false
        monthly_limit_bytes    = 104857600
```

I'd expect that the state contains the same values as when a distribution is created without these two values in the config:

```
default_cache_duration = "P1D"
```
and `monthly_limit_bytes` missing completely from state.

File: stackit/internal/services/cdn/distribution/resource.go [hunk 8]
```diff
@@ -1446,10 +1489,40 @@ func mapFields(ctx context.Context, distribution *cdnSdk.Distribution, model *Mo
...
+	if diags.HasError() {
+		return core.DiagsToError(diags)
+	}
+
+	var defaultCacheDuration types.String
+	if distribution.Config.DefaultCacheDuration.IsSet() {
+		defaultCacheDuration = types.StringPointerValue(distribution.Config.DefaultCacheDuration.Get())
+	} else {
+		defaultCacheDuration = types.StringNull()
+	}
```
Comment (cgoetz-inovex) [inline stackit/internal/services/cdn/distribution/resource.go]: Nitpick: could be simplified
```suggestion
	defaultCacheDuration := types.StringPointerValue(distribution.Config.DefultCacheDuration.Get())
```
These NullableStringTypes are a bit unintuitive, `Get()` is always save to call, it will just return nil. When passed to `types.StringPointerValue()` this would result in a `StringNull`:
```go
func TestNullableStringToTerraformString(t *testing.T) {
	uninitialized := cdnSdk.NullableString{}

	setNil := cdnSdk.NullableString{}
	setNil.Set(nil)

	uninitializedValue := types.StringPointerValue(uninitialized.Get())
	setNilValue := types.StringPointerValue(setNil.Get())

	if diff := cmp.Diff(types.StringNull(), uninitializedValue); diff != "" {
		t.Fatalf("uninitialized NullableString converted unexpectedly (-want +got):\n%s", diff)
	}
	if diff := cmp.Diff(uninitializedValue, setNilValue); diff != "" {
		t.Fatalf("uninitialized and nil NullableString conversions differ (-uninitialized +nil):\n%s", diff)
	}
}
```

File: stackit/internal/services/cdn/distribution/resource.go [hunk 9]
```diff
@@ -1446,10 +1489,40 @@ func mapFields(ctx context.Context, distribution *cdnSdk.Distribution, model *Mo
...
+
+	var monthlyLimitBytes types.Int64
+	if distribution.Config.MonthlyLimitBytes.IsSet() {
+		monthlyLimitBytes = types.Int64PointerValue(distribution.Config.MonthlyLimitBytes.Get())
+	} else {
+		monthlyLimitBytes = types.Int64Null()
+	}
```
Comment (cgoetz-inovex) [inline stackit/internal/services/cdn/distribution/resource.go]: same as above

File: stackit/internal/services/cdn/distribution/datasource.go [hunk 10]
```diff
@@ -678,11 +694,41 @@ func mapDataSourceFields(ctx context.Context, distribution *cdnSdk.Distribution,
...
+	} else {
+		defaultCacheDuration = types.StringNull()
+	}
+
+	var monthlyLimitBytes types.Int64
+	if distribution.Config.MonthlyLimitBytes.IsSet() {
+		monthlyLimitBytes = types.Int64PointerValue(distribution.Config.MonthlyLimitBytes.Get())
+	} else {
+		monthlyLimitBytes = types.Int64Null()
+	}
```
Comment (cgoetz-inovex) [inline stackit/internal/services/cdn/distribution/datasource.go:722]: see resource for nitpick suggestion, but these aren't a merge blocker from my side

File: stackit/internal/services/cdn/distribution/resource.go (hunk 7 above)
Comment (matheuspolitano) [reply stackit/internal/services/cdn/distribution/resource.go:1037]: This is computed value if you remove will keep the lastest value we are reflecting our api behavior

File: stackit/internal/services/cdn/distribution/datasource_test.go (hunk 1 above)
Comment (cgoetz-inovex) [reply stackit/internal/services/cdn/distribution/datasource_test.go:69]: I think the commit where this was done is missing from the PR

File: stackit/internal/services/cdn/distribution/resource.go (hunk 7 above)
Comment (matheuspolitano) [reply stackit/internal/services/cdn/distribution/resource.go:1037]: I added this information in our docs

File: stackit/internal/services/cdn/distribution/resource.go (hunk 9 above)
Comment (matheuspolitano) [reply stackit/internal/services/cdn/distribution/resource.go]: Done

File: stackit/internal/services/cdn/distribution/datasource.go (hunk 10 above)
Comment (matheuspolitano) [reply stackit/internal/services/cdn/distribution/datasource.go:722]: Done

File: stackit/internal/services/cdn/distribution/resource.go (hunk 8 above)
Comment (matheuspolitano) [reply stackit/internal/services/cdn/distribution/resource.go]: Done

File: stackit/internal/services/cdn/distribution/datasource_test.go (hunk 1 above)
Comment (cgoetz-inovex) [reply stackit/internal/services/cdn/distribution/datasource_test.go:69]: resolving, this is covered in acc tests

**Review by cgoetz-inovex — APPROVED** [review]
